src/atomic.rs

```diff
@@ -0,0 +1,878 @@
+        thread::spawn(move || assert!(sa.listen().is_err()));
+        thread::spawn(move || assert!(sb.listen().is_err()));
+        thread::spawn(move || assert!(sc.listen().is_err()));
+
+        let mut primary = Primary::new(payloads(&a, &b, &c), bus(), None, timeout(5), None);
+        assert!(primary.commit().is_err());
+        assert_eq!(primary.committed(), &hashset!{});
+        assert_eq!(primary.aborted(), &hashset!{a, b});
+    }
+
+    #[test]
+    fn atomic_prepare_timeout() {
+        let (a, b, c) = serials("prepare_timeout");
+        let mut sa = Secondary::new(a.clone(), bus(), Box::new(Success), timeout(1), None);
+        let mut sb = Secondary::new(b.clone(), bus(), Box::new(Success), timeout(1), None);
+        let mut sc = Secondary::new(c.clone(), bus(), Box::new(PrepareTimeout), timeout(1), None);
+        thread::spawn(move || assert!(sa.listen().is_err()));
+        thread::spawn(move || assert!(sb.listen().is_err()));
+        thread::spawn(move || assert!(sc.listen().is_err()));
+
+        let mut primary = Primary::new(payloads(&a, &b, &c), bus(), None, timeout(5), None);
+        assert!(primary.commit().is_err());
+        assert_eq!(primary.committed(), &hashset!{});
+        assert_eq!(primary.aborted(), &hashset!{a, b});
+    }
+
+    #[test]
+    fn atomic_commit_timeout() {
+        let (a, b, c) = serials("commit_timeout");
+        let mut sa = Secondary::new(a.clone(), bus(), Box::new(Success), timeout(1), None);
+        let mut sb = Secondary::new(b.clone(), bus(), Box::new(Success), timeout(1), None);
+        let mut sc = Secondary::new(c.clone(), bus(), Box::new(CommitTimeout), timeout(1), None);
+        thread::spawn(move || assert!(sa.listen().is_ok()));
+        thread::spawn(move || assert!(sb.listen().is_ok()));
+        thread::spawn(move || assert!(sc.listen().is_ok()));
+
+        let mut primary = Primary::new(payloads(&a, &b, &c), bus(), None, timeout(5), None);
+        assert!(primary.commit().is_err());
+        assert_eq!(primary.committed(), &hashset!{a, b});
+        assert_eq!(primary.aborted(), &hashset!{});
+    }
+
+    #[cfg(not(feature = "docker"))]
+    #[test]
+    fn atomic_verify_crash() {
+        let (a, b, c) = serials("verify_crash");
+        let c2 = c.clone();
+        let mut sa = Secondary::new(a.clone(), bus(), Box::new(Success), timeout(9), None);
+        let mut sb = Secondary::new(b.clone(), bus(), Box::new(Success), timeout(9), None);
+        thread::spawn(move || assert!(sa.listen().is_ok()));
+        thread::spawn(move || assert!(sb.listen().is_ok()));
+        thread::spawn(move || {
+            let path = format!("/tmp/sota-atomic-verify-crash-{}", time::precise_time_ns().to_string());
+            let outcome = panic::catch_unwind(|| {
+                let mut sc = Secondary::new(c2, bus(), Box::new(VerifyCrash), timeout(9), Some(path.clone()));
+                panic::set_hook(Box::new(|_| ()));
+                assert!(sc.listen().is_err());
```

Comment:
So if the panic does not happen, `assert!(outcome.is_err())` will fail and the successful case will never happen, unless I have misunderstood?